File: src/Itinero/Osm/Vehicles/Moped.cs
// Itinero - Routing for .NET
// Copyright (C) 2016 Abelshausen Ben
// 
// This file is part of Itinero.
// 
// Itinero is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 2 of the License, or
// (at your option) any later version.
// 
// Itinero is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with Itinero. If not, see <http://www.gnu.org/licenses/>.

using Itinero.Attributes;
using Itinero.Profiles;
using System.Collections.Generic;

namespace Itinero.Osm.Vehicles
{

    /// <summary>
    /// Represents the default OSM moped profile.
    /// </summary>
    public class Moped : Vehicle
    {
        /// <summary>
        /// Gets the name of this vehicle.
        /// </summary>
        public override string Name
        {
            get
            {
                return "Moped";
            }
        }

        /// <summary>
        /// Gets the vehicle types.
        /// </summary>
        public override string[] VehicleTypes
        {
            get
            {
                return new string[] { "vehicle", "motor_vehicle", "moped" };
            }
        }

        /// <summary>
        /// Gets a whitelist of attributes to keep as meta-data.
        /// </summary>
        public override HashSet<string> MetaWhiteList
        {
            get
            {
                return new HashSet<string>(new[] { "name" });
            }
        }

        /// <summary>
        /// Gets a whitelist of attributes to keep as part of the profile.
        /// </summary>
        public override HashSet<string> ProfileWhiteList
        {
            get
            {
                return new HashSet<string>();
            }
        }

        /// <summary>
        /// Get a function to calculate properties for a set given edge attributes.
        /// </summary>
        /// <returns></returns>
        public sealed override FactorAndSpeed FactorAndSpeed(IAttributeCollection attributes, Whitelist whiteList)
        {
            string highway = string.Empty;
            if (attributes == null ||
                !attributes.TryGetValue("highway", out highway))
            {
                return Profiles.FactorAndSpeed.NoFactor;
            }

            var speed = 70.0f;
            var canstopon = true;
            switch (highway)
            {
                case "services":
                case "living_street":
                    speed = 5;
                    break;
                case "service":
                case "track":
                case "road":
                case "residential":
                case "unclassified":
                case "tertiary":
                case "tertiary_link":
                case "secondary":
                case "secondary_link":
                case "trunk":
                case "trunk_link":
                case "primary":
                case "primary_link":
                    speed = 40;
                    break;
                default:
                    return Profiles.FactorAndSpeed.NoFactor;
            }
            whiteList.Add("highway");

            // get max-speed tag if any.
            var maxSpeed = 0.0f;
            if (attributes.TryGetMaxSpeed(out maxSpeed))
            {
                whiteList.Add("maxspeed");
                if (speed > 40)
                {
                    speed = 40;
                }
                else
                {
                    speed = maxSpeed * 0.75f;
                }
            }

            // access tags.
            if (!Vehicle.InterpretAccessValues(attributes, whiteList, this.VehicleTypes, "access"))
            {
                return Profiles.FactorAndSpeed.NoFactor;
            }

            // oneway restrictions.
            short direction = 0; // 0=bidirectional, 1=forward, 2=backward
            string oneway;
            string junction;
            if (attributes.TryGetValue("junction", out junction))
            {
                if (junction == "roundabout")
                {
                    whiteList.Add("junction");
                    direction = 1;
                }
            }
            if (attributes.TryGetValue("oneway", out oneway))
            {
                if (oneway == "yes")
                {
                    direction = 1;
                }
                else if (oneway == "no")
                { // explicitly tagged as not oneway.

                }
                else
                {
                    direction = 2;
                }
                whiteList.Add("oneway");
            }

            speed = speed / 3.6f; // to m/s
            if (!canstopon)
            { // add canstop on info to direction.
                direction += 3;
            }

            return new Profiles.FactorAndSpeed()
            {
                Constraints = null,
                Direction = direction,
                SpeedFactor = 1.0f / speed,
                Value = 1.0f / speed
            };
        }
    }
}

File: src/Itinero/Osm/Vehicles/Lua/moped.lua
-- moped globals
name = "moped"
vehicle_types = { "vehicle", "motor_vehicle", "moped" }

minspeed = 40
maxspeed = 40

-- default speed profiles
speed_profile = {
	["primary"] = 40,
	["primary_link"] = 40,
	["secondary"] = 40,
	["secondary_link"] = 40,
	["tertiary"] = 40,
	["tertiary_link"] = 40,
	["unclassified"] = 40,
	["residential"] = 40,
	["service"] = 30,
	["services"] = 30,
	["road"] = 30,
	["track"] = 30,
	["living_street"] = 5,
	["ferry"] = 5,
	["movable"] = 5,
	["shuttle_train"] = 10,
  	["default"] = 10
}

-- default access values
access_values = {
	["private"] = false,
	["yes"] = true,
	["no"] = false,
	["permissive"] = true,
	["destination"] = true,
	["customers"] = false,
	["designated"] = true,
	["public"] = true,
	["delivery"] = true,
	["use_sidepath"] = false
}

-- whitelists for profile and meta
profile_whitelist = {

}
meta_whitelist = {
	"name"
}

-- profile definitions linking a function to a profile
profiles = {
	{
		name = "",
		function_name = "factor_and_speed",
		metric = "time"
	},
	{ 
		name = "shortest",
		function_name = "factor_and_speed",
		metric = "distance",
	}
}

-- interprets access tags
function can_access (attributes, result)
	local last_access = nil
	local access = access_values[attributes.access]
	if access != nil then
		result.attributes_to_keep.access = true
		last_access = access
	end
	for i=0, 10 do
		local access_key_key = vehicle_types[i]
		local access_key = attributes[access_key_key]
		if access_key then
			access = access_values[access_key]
			if access != nil then
				result.attributes_to_keep[access_key_key] = true
				last_access = access
			end
		end
	end
	return last_access
end

-- turns a oneway tag value into a direction
function is_oneway (attributes, name)
	local oneway = attributes[name]
	if oneway != nil then
		if oneway == "yes" or 
		   oneway == "true" or 
		   oneway == "1" then
			return 1
		end
		if oneway == "-1" then
			return 2
		end
	end
	return nil
end

-- the main function turning attributes into a factor_and_speed and a tag whitelist
function factor_and_speed (attributes, result)
	 local highway = attributes.highway
	 
	 result.speed = 0
	 result.direction = 0
	 result.canstop = true
	 result.attributes_to_keep = {}

	 -- get default speed profiles
	 local highway_speed = speed_profile[highway]
	 if highway_speed then
        result.speed = highway_speed
        result.direction = 0
		result.canstop = true
		result.attributes_to_keep.highway = highway
		if highway == "motorway" or 
		   highway == "motorway_link" then
		   result.canstop = false
		end
	 else
	    return
	 end

	 -- interpret access tags
	 if can_access (attributes, result) == false then
	 	result.speed = 0
		result.direction = 0
		result.canstop = true
	    return
	 end
	 
	 -- get maxspeed if any.
	 if attributes.maxspeed then
		local speed = itinero.parsespeed (attributes.maxspeed)
		if speed then
			if speed > 40 then
				speed = 40
			end
			result.speed = speed * 0.75
			result.attributes_to_keep.maxspeed = true
		end
	end

	-- get directional information
	local junction = attributes.junction
	if junction == "roundabout" then
		result.direction = 1
		result.attributes_to_keep.junction = true
	end
	local direction = is_oneway (attributes, "oneway")
	if direction != nil then
		result.direction = direction
		result.attributes_to_keep.oneway = true
	end
end